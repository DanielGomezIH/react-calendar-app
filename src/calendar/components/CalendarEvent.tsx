import { Event } from '@/models';

export const CalendarEvent = ( { event } ) => {

  const { title, user }: Event = event;

  return (
    <>
      <strong>{ title }</strong>
      <span> - { user.name }</span>
    </>
  );
};