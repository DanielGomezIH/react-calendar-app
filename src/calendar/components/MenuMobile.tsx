import { Link } from 'react-router-dom';
import { Button } from '@/components/ui/button';
import { Sheet, SheetContent, SheetTrigger } from '@/components/ui/sheet';
import { Calendar, Menu } from 'lucide-react';

export const MenuMobile = () => {
  return (
    <Sheet>
      <SheetTrigger asChild>
        <Button
          variant="outline"
          size="icon"
          className="shrink-0 md:hidden"
        >
          <Menu className="h-5 w-5" />
          <span className="sr-only">Men√∫ de usuario</span>
        </Button>
      </SheetTrigger>

      <SheetContent side="left">
        <nav className="grid gap-6 text-lg font-medium">
          <Link
            to="/"
            className="flex items-center gap-2 text-lg font-semibold"
          >
            <Calendar className="h-6 w-6" />
            <span className="sr-only">Calendar App</span>
          </Link>

          <Link to="/" className="hover:text-foreground">
            Dashboard
          </Link>

          <Link
            to="/"
            className="text-muted-foreground hover:text-foreground"
          >
            Orders
          </Link>

          <Link
            to="/"
            className="text-muted-foreground hover:text-foreground"
          >
            Products
          </Link>
        </nav>
      </SheetContent>
    </Sheet>
  );
};