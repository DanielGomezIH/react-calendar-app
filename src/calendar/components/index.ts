export * from './CalendarEvent';
export * from './NewEventForm';
export * from './NewEventModal';
export * from './AddNewButton.tsx';
