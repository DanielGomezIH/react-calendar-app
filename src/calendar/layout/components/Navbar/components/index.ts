export * from './MenuMobile';
export * from './MenuUser';
export * from './MenuWeb';
