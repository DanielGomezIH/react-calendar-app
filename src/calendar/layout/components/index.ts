export * from './MenuMobile';
export * from './MenuUser';
export * from './MenuWeb';
export * from './Navbar';
