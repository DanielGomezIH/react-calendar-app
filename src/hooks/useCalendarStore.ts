import { Event } from '@/models';
import {
  onAddNewEvent,
  onDeleteEvent,
  onDeselectEvent,
  onSelectEvent,
  onSetActiveEvent,
  onUpdateEvent,
  useAppDispatch,
  useAppSelector,
} from '@/store';

export const useCalendarStore = () => {
  const dispatch = useAppDispatch();
  const { events, activeEvent, hasEventSelected } = useAppSelector(
    (state) => state.calendar,
  );

  const setActiveEvent = (calendarEvent: Event) => {
    dispatch(onSetActiveEvent(calendarEvent));
  };

  const startSavingEvent = async (calendarEvent: Event) => {
    if (calendarEvent._id !== null) {
      dispatch(onUpdateEvent({ ...calendarEvent }));
    } else {
      dispatch(
        onAddNewEvent({
          ...calendarEvent,
          _id: new Date().getTime(),
        }),
      );
    }
  };

  const startDeletingEvent = async (calendarEvent: Event) => {
    dispatch(onDeleteEvent({ ...calendarEvent }));
  };

  const setHasSelectedEvent = () => {
    dispatch(onSelectEvent());
  };

  const setHasNotSelectedEvent = () => {
    dispatch(onDeselectEvent());
  };

  return {
    //Properties
    events,
    activeEvent,
    hasEventSelected,

    //Methods
    setActiveEvent,
    startSavingEvent,
    startDeletingEvent,
    setHasSelectedEvent,
    setHasNotSelectedEvent,
  };
};
