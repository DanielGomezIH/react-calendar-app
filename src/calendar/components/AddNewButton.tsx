import { Button } from '@/components/ui/button';
import { useCalendarStore, useUiStore } from '@/hooks';
import { addHours } from 'date-fns';
import { Plus } from 'lucide-react';

export const AddNewButton = () => {

  const { openDateModal } = useUiStore();
  const { setActiveEvent } = useCalendarStore();

  const onClickNew = () => {
    setActiveEvent( {
      _id: null,
      start: new Date(),
      end: addHours( new Date(), 2 ),
      title: '',
      notes: '',
      bgColor: '#fafafa',
      user: {
        _id: 123,
        name: 'Daniel',
      },
    } );
    openDateModal();
  };

  return (
    <Button
      className="fixed rounded-full bottom-6 right-6 h-auto w-auto p-2"
      onClick={ onClickNew }
    >
      <Plus size={ 20 } />
    </Button>
  );
};