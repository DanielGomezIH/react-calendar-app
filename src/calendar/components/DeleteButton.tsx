import { Button } from '@/components/ui/button';
import { useCalendarStore, useUiStore } from '@/hooks';
import { Trash } from 'lucide-react';

export const DeleteButton = () => {

  const { isDateModalOpen } = useUiStore();
  const { activeEvent, startDeletingEvent } = useCalendarStore();

  const showDeleteButton = activeEvent !== null && activeEvent.id !== null && isDateModalOpen === false;

  const onClickNew = () => {
    startDeletingEvent( activeEvent! );
  };

  return (
    <>
      { showDeleteButton && (
        <Button
          variant='destructive'
          className="fixed rounded-full bottom-[70px] right-6 h-auto w-auto p-2"
          onClick={ onClickNew }
        >
          <Trash size={ 20 } />
        </Button>
      ) }
    </>
  );
};