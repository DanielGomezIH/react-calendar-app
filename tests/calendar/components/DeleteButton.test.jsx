import { render, screen, fireEvent } from '@testing-library/react';

import { DeleteButton } from '../../../src/calendar/components/DeleteButton';

import { useCalendarStore } from '../../../src/hooks/useCalendarStore.ts';
import { useUiStore } from '../../../src/hooks/useUiStore.ts';

import { events } from '../../fixtures/calendarStates.ts';

jest.mock( '../../../src/hooks/useCalendarStore.ts' );
jest.mock( '../../../src/hooks/useUiStore.ts' );

describe( 'Pruebas en <DeleteButton />', () => {

  const mockStartDeletingEvent = jest.fn();

  beforeEach( () => jest.clearAllMocks() );

  test( 'Debe de mostrar el componente correctamente', () => {

    useCalendarStore.mockReturnValue( {
      activeEvent: { ...events[ 0 ] }
    } );

    useUiStore.mockReturnValue( {
      isDateModalOpen: false
    } );

    const { container } = render( <DeleteButton /> );

    const buttonHTML = screen.findByRole( 'button' );

    expect( buttonHTML ).toBeTruthy();

    expect( container ).toMatchSnapshot();

  } );

  test( 'No debe mostrar el componente si no hay un evento activo', () => {

    useCalendarStore.mockReturnValue( {
      activeEvent: null
    } );

    useUiStore.mockReturnValue( {
      isDateModalOpen: false
    } );

    const { container } = render( <DeleteButton /> );

    expect( container ).toMatchSnapshot();

  } );


  test( 'Debe de llamar startDeletingEvent si hay un evento activo', async () => {

    useCalendarStore.mockReturnValue( {
      activeEvent: { ...events[ 0 ] },
      startDeletingEvent: mockStartDeletingEvent
    } );

    useUiStore.mockReturnValue( {
      isDateModalOpen: false
    } );

    render( <DeleteButton /> );

    const buttonHTML = await screen.findByRole( 'button' );

    fireEvent.click( buttonHTML );

    expect( mockStartDeletingEvent ).toHaveBeenCalled();

  } );

} );
