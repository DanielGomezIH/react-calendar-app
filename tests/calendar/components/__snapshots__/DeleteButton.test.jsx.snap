// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Pruebas en <DeleteButton /> Debe de mostrar el componente correctamente 1`] = `
<div>
  <button
    class="inline-flex items-center justify-center whitespace-nowrap text-sm font-medium ring-offset-white transition-colors focus-visible:outline-none focus-visible:ring-2 focus-visible:ring-navy-950 focus-visible:ring-offset-2 disabled:pointer-events-none disabled:opacity-50 dark:ring-offset-navy-950 dark:focus-visible:ring-navy-300 bg-red-500 text-navy-50 hover:bg-red-500/90 dark:bg-red-900 dark:text-navy-50 dark:hover:bg-red-900/90 fixed rounded-full bottom-[70px] right-6 h-auto w-auto p-2"
  >
    <svg
      class="lucide lucide-trash"
      fill="none"
      height="20"
      stroke="currentColor"
      stroke-linecap="round"
      stroke-linejoin="round"
      stroke-width="2"
      viewBox="0 0 24 24"
      width="20"
      xmlns="http://www.w3.org/2000/svg"
    >
      <path
        d="M3 6h18"
      />
      <path
        d="M19 6v14c0 1-1 2-2 2H7c-1 0-2-1-2-2V6"
      />
      <path
        d="M8 6V4c0-1 1-2 2-2h4c1 0 2 1 2 2v2"
      />
    </svg>
  </button>
</div>
`;

exports[`Pruebas en <DeleteButton /> No debe mostrar el componente si no hay un evento activo 1`] = `<div />`;
